src/Components/Web.JS/src/Rendering/DomMerging/DataPermanentElementSync.ts

```diff
@@ -0,0 +1,23 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+export function isDataPermanentElement(elem: Element): boolean {
+  return elem.hasAttribute('data-permanent');
+}
+
+export function areIncompatibleDataPermanentElements(elementA: Element, elementB: Element) {
+  const isDataPermanentA = isDataPermanentElement(elementA);
+  const isDataPermanentB = isDataPermanentElement(elementB);
+
+  if (isDataPermanentA !== isDataPermanentB) {
+    // A 'data permanent' element can't be merged with a 'non-data-permanent' one.
+    return true;
+  }
+
+  if (isDataPermanentA && elementA.id !== elementB.id) {
```

Comment:
This logic doesn't account for scenarios like `<div data-permanent>`, or at least would treat that as *not* permanent since the empty string attribute value would be considered falsey on line 17. That could be really confusing. Since `data-enhance` allows valueless attributes, it would be good to do that here too.

Also, I thought we were changing the rule not to use `id`, but instead use the values of the `data-permanent` attributes to control this.

I'd also be happy with some rule like:

 * If `data-permanent` is present:
   * If its value is an empty-string (like it is for valueless attributes)
     * If the element has an ID, use that as the data permanence value
     * Else use the fixed string "true" as the data permanence value
   * Else use the attribute value as the data permanence value

TBH I'd also be happy with *not* using `element.id` at all, since it's yet another avoidable thing to document and support forever. This is based on the diff algorithm not matching on a global basis, and so it generally is going to behave as desired in many cases if you simply have `data-permanent` with no attribute value.

What do you think?

Reply:
I think relying on the `data-permanent` attribute value and nothing else (ignoring `element.id`) sounds good. I've just updated the PR accordingly.